### ISSUE TITLE
Adding a close() method

### ISSUE THREAD

**Issue by EliFinkelshteyn** [twmht/python-rocksdb#10]
This a request that came up [many](https://github.com/stephan-hof/pyrocksdb/issues/10) [times](https://github.com/stephan-hof/pyrocksdb/issues/32) on the old pyrocksdb repo, but was never implemented. The reasoning seemed off to me. Without a `close()` method, there really is no way to guarantee a lock is released as Python's garbage collection can't fully be relied on for this. 

If you're consistently recreating databases from scratch (as I am in my use-case), this really complicates things. Any chance such a method can be added?

**Comment by maxux** [comment]
I hit the same issue.
I need to create a db from scratch and I there is no way to know or wait for the database to reach a stable state. A blocking `close()` method would solve this easily.

**Comment by maxux** [comment]
A workaround could be using `.compact_range()` as replacement of `.close()` which helps to ensure stability.

**Comment by EliFinkelshteyn** [comment]
@maxux I wound up defining a close method as follows:

```python
def close(self):
    del self._db
    gc.collect()
```

This is essentially what's done in the core rocks db library for their tests as well. I don't love relying on `gc.collect()`, but we haven't had any issues with it here yet. I'd lean away from using `.compact_range()` because it can be slow, especially as the db grows.

**Comment by ask** [comment]
I'd say this is a pretty important issue for anybody who require graceful shutdown of processes.  Relying on __del__ is a horrible practice that should be avoided if possible, and IMHO if necessary
even warrants a backward incompatible change.

Doing `del(self._db)` may work for that specific application, but in general you'd have to carefully
del() any other references to the object as well, and who knows what happens if there are circular references.

NOTE: Using a weakref callback you could possibly set a flag to ensure the database was closed
(well, implicitly by side effect of being garbage collected, trying to be an optimist here:)
pseudocode:


```py
class RocksDBWrapper:
    _db_collected = False

    def open(self):
        self._db = rocksdb.DB(...)
        self._dbref = weakref.ref(self._db, self._on_db_collected)
       
    def _on_db_collected(self, ref):
        self._db_collected = True

    def close(self):
        while not self._db_collected:
            try:
                del(self.db)
            except AttributeError:
                pass
            gc.collect()
```

I think this is a pretty serious issue that needs to be resolved. Hopefully I, or someone else get some time to look into it.

**Comment by EliFinkelshteyn** [comment]
Can't agree with @ask more. I can try to make some time to help build or test, if you'd like a hand. We're a bit overworked at work currently, but this is an issue near and dear to my heart, so happy to make some time over the weekend.

**Comment by fish2000** [comment]
+1 for @ask’s context-manager suggestion.

I have two PRs in for a `db.close()` method, #39 (which can no longer be merged due to conflicts) and #40 (whose tests are failing due to a separate, unrelated change). Stay tuned!

**Comment by twmht** [comment]
closed due to fix by #40

**Comment by entangledloops** [comment]
Looks like there was never a push to pypi after this merge, so this feature is still not available if using `pip install python-rocksdb`.
https://pypi.org/project/python-rocksdb/0.7.0/#history
